Fix deadlock in LDAP assigned role update

jmaicher opened the pull request:
`LDAPAccessStorage` performs role lookups via `AccessControl::find` in `update_func` passed to `MemoryAccessStorage::update`, which is guarded by a mutex. This can lead to deadlocks for non-existing roles caused by reentrance via `MultipleAccessStorage::find` and `LDAPAccessStorage::find` to `MemoryAccessStorage::find`, which is guarded by the same mutex.

The provided integration test reproduces the problem and passes with the given changes. 

Note:
- This is a regression and was already fixed once with https://github.com/ClickHouse/ClickHouse/pull/24431
- We already use `std::recursive_mutex` for the same reason in `LDAPAccessStorage`. 

Fixes #54318, #53708, #41380

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix deadlock in LDAP assigned role update for non-existing ClickHouse roles

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=55119) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=55119) before we can accept your contribution.<br/><hr/>**Julian Maicher** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=55119) it.</sub>

jmaicher committed:
test: Add IT that reproduces deadlock

jmaicher committed:
fix: Use recursive mutex to prevent deadlock

`LDAPAccessStorage` performs role lookups via `AccessControl::find` in `update_func` passed to `MemoryAccessStorage::update`, which is guarded by a mutex.
This can lead to deadlocks for non-existing roles caused by reentrance via `MultipleAccessStorage::find` and `LDAPAccessStorage::find` to `MemoryAccessStorage::find`, which is guarded by the same mutex.

Note: We already use `std::recursive_mutex` for the same reason in `LDAPAccessStorage`.

Fixes #54318

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #55119 from jmaicher/ClickHouse:fix/54318/deadlock-ldap-role-update -->
*This is an automated comment for commit f2b8566df0eb0d8dc29836163a8cc7d51cd91700 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55119/f2b8566df0eb0d8dc29836163a8cc7d51cd91700/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

jmaicher committed:
style: Apply black formatter

jmaicher commented:
Can we rerun the failed checks? I suspect they are flaky.

jmaicher commented:
Can we backport the fix to 23.8 LTS?

lavasser commented:
And can we backport the fix to 23.3 LTS?

tavplubix commented:
@jmaicher, the new test is flaky: 
https://s3.amazonaws.com/clickhouse-test-reports/0/5a830834fe564ed91c2dceb7b28973d5a70c0858/integration_tests__release__[4_4].html
https://s3.amazonaws.com/clickhouse-test-reports/0/d9a18faab8ebb23c1d3daaca66203137a30459e8/integration_tests__release__[4_4].html

jmaicher commented:
I'll have a look.

mark-bb commented:
Hello.
Still no backport of the fix to 23.3.
Is it planned?

tavplubix commented:
@jmaicher, the new test is still flaky: https://s3.amazonaws.com/clickhouse-test-reports/55540/517ce45df7f52f970f9d9f12fc606b2b0fc62b0d/integration_tests__release__[4_4].html

jmaicher commented:
@tavplubix I know, haven't been able to reproduce it locally and got side-tracked. I will figure it out.

jmaicher commented:
@tavplubix Hopefully fixed with https://github.com/ClickHouse/ClickHouse/pull/56544. I wasn't able to reproduce it, but based on all the logs I'm pretty sure it's a DNS issue.
